store/index.go

```diff
@@ -0,0 +1,324 @@
+func indexDelete(indexer Indexer, tx ds.Txn, key ds.Key, originalData []byte) error {
+	indexes := indexer.Indexes()
+
+	for path, index := range indexes {
+		err := indexUpdate(indexer.BaseKey(), path, index, tx, key, originalData, true)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// adds or removes a specific index on an item
+func indexUpdate(baseKey ds.Key, path string, index Index, tx ds.Txn, key ds.Key, value []byte,
+	delete bool) error {
+
+	valueKey, err := index.IndexFunc(path, value)
+	if valueKey.String() == "" {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	indexKey := indexPrefix.Child(baseKey).ChildString(path).Instance(valueKey.String()[1:])
+
+	data, err := tx.Get(indexKey)
+	if err != nil && err != ds.ErrNotFound {
+		return err
+	}
+
+	indexValue := make(keyList, 0)
+
+	if err != ds.ErrNotFound {
+		if index.Unique && !delete {
+			return ErrUniqueExists
+		}
+	}
+
+	if data != nil {
+		err = DefaultDecode(data, &indexValue)
+		if err != nil {
+			return err
+		}
+	}
+
+	if delete {
+		indexValue.remove(key)
+	} else {
+		indexValue.add(key)
+	}
+
+	if len(indexValue) == 0 {
+		return tx.Delete(indexKey)
+	}
+
+	iVal, err := DefaultEncode(indexValue)
+	if err != nil {
+		return err
+	}
+
+	return tx.Put(indexKey, iVal)
+}
+
+// keyList is a slice of unique, sorted keys([]byte) such as what an index points to
+type keyList [][]byte
+
+func (v *keyList) add(key ds.Key) {
+	b := key.Bytes()
+	i := sort.Search(len(*v), func(i int) bool {
+		return bytes.Compare((*v)[i], b) >= 0
+	})
+
+	if i < len(*v) && bytes.Equal((*v)[i], b) {
+		// already added
+		return
+	}
+
+	*v = append(*v, nil)
+	copy((*v)[i+1:], (*v)[i:])
+	(*v)[i] = b
+}
+
+func (v *keyList) remove(key ds.Key) {
+	b := key.Bytes()
+	i := sort.Search(len(*v), func(i int) bool {
+		return bytes.Compare((*v)[i], b) >= 0
+	})
+
+	if i < len(*v) {
+		copy((*v)[i:], (*v)[i+1:])
+		(*v)[len(*v)-1] = nil
+		*v = (*v)[:len(*v)-1]
+	}
+}
+
+func (v *keyList) in(key ds.Key) bool {
+	b := key.Bytes()
+	i := sort.Search(len(*v), func(i int) bool {
+		return bytes.Compare((*v)[i], b) >= 0
+	})
+
+	return (i < len(*v) && bytes.Equal((*v)[i], b))
+}
+
+// FilterQuery provides a Query Filter function for querying JSON data
+type FilterQuery struct {
+	query *JSONQuery
+}
+
+// Filter is the primary Filter function
+func (f FilterQuery) Filter(e query.Entry) bool {
+	val := make(map[string]interface{})
+	if err := json.Unmarshal(e.Value, &val); err != nil {
+		return false
+	}
+	ok, err := f.query.matchJSON(val)
+	return ok && err == nil
+}
+
+type iterator struct {
+	nextKeys func() ([]ds.Key, error)
+	txn      ds.Txn
+	query    *JSONQuery
+	err      error
+	keyCache []ds.Key
+	iter     query.Results
+}
+
+func newIterator(txn ds.Txn, baseKey ds.Key, q *JSONQuery) *iterator {
+	i := &iterator{
+		txn:   txn,
+		query: q,
+	}
+	// Key field or index not specified, pass thru to base 'iterator'
+	if q.Index == "" {
```

Comment:
If we didn't specify an index, we just essentially wrap the default query.Query and return the results that way. That way, not using an index has no real impact on our query from how things were before.